### PR TITLE
Bug 2034364: Use resource{apply,merge,read} functions provided by library-go

### PR DESCRIPTION
**Pull request by mkenigs**
Borrowing from https://github.com/openshift/machine-config-operator/pull/829 which helps https://github.com/openshift/machine-config-operator/issues/819

We currently have code duplicated from library-go in ./lib, and this
commit uses some of the library-go functions instead and removes the
files that contained the duplicated code in
./lib/resource{apply,merge,read}.

Doing so requires adding events.Recorder as libgoRecorder to the
Operator struct.

Note that this does not remove all functions from lib; some lib
functions contain MCO specific logic or would not be trivial to move, so
doing so can be done at another time.

Unit testing these changes is not really possible since the newly used
functions are from outside the MCO repo, and the complexity of e2e
testing is not worth the effort at this time, since the library-go
functions should maintain current behavior. A more thorough test plan
might be necessary to move the rest of lib.

### CODE REVIEW

**Comment by mkenigs** [comment]
@kikisdeliveryservice how would I go about verifying this doesn't break anything? Looked at some of the functions in `lib` and there are definitely differences, but I don't know if they're important

**Comment by yuqi-zhang** [comment]
This one is a bit harder to test since we never really had the full scope of unit testing for these libraries. We could attempt to add MCO specific field testing for this, though, just to make sure we don't break anything.

Two ways we could approach this:
1. look at the old functions (e.g. what we are ensuring when updating the daemonsets, etc.) and attempt to mock changing those via unit tests
2. do some manual testing by modifying fields in a live cluster

File: pkg/operator/sync_test.go
Comment (mkenigs) [inline pkg/operator/sync_test.go:169]: How would I get a mock `apiExtClient`?

File: pkg/operator/sync_test.go
Comment (mkenigs) [inline pkg/operator/sync_test.go:182]: @yuqi-zhang when you look at what we are ensuring with the old functions, do I need to make more specific assertions about what `syncCustomResourceDefinitions()` should be doing here?

Comment (yuqi-zhang) [reply pkg/operator/sync_test.go:182]: Answered in chat, but to also summarize here, we should aim to test with mock definitions of what fields we care about in the various MCO resource definitions, to ensure creation/updates/deletion is handled correctly

**Comment by kikisdeliveryservice** [comment]
depends on: #2837

**Review by kikisdeliveryservice — APPROVED** [review]
This is in good shape, defer to Jerry for final review

/assign @yuqi-zhang

**Comment by sinnykumari** [comment]
Most of the tests are failing. To avoid un-necessary retest putting hold until they are fixed
/hold

Feel free to  remove hold once test looks good

**Comment by sinnykumari** [comment]
Need fix, most of the tests are failing

**Comment by mkenigs** [comment]
/hold cancel
Somehow I had missed one of the ApplySecret functions

**Review by yuqi-zhang — APPROVED** [review]
/lgtm

updated function with context looks good

**Comment by sinnykumari** [comment]
skipping optional test to save bot keeping retesting
/skip

**Comment by sinnykumari** [comment]
Putting hold until bugzilla and commit message is updated with with bug fix this PR does.
/hold

**Comment by yuqi-zhang** [comment]
Sorry for the delay, updated bugzilla description. If that looks good, feel free to remove hold. Thanks!

**Comment by sinnykumari** [comment]
Thanks Jerry!
/unhold

**Pull request merged**
